exec/Miner/Updates.hs

```diff
@@ -72,38 +67,39 @@ getUpdateVar (UpdateMap v) k = modifyMVar v $ \m -> case HM.lookup k m of
     newUpdateStream var = T2 var
         <$!> async (updateStream (_updateKeyChainId k) var)
 
-    -- TODO:
-    --
-    -- We don't reap old entries from the map. That's fine since the maximum
-    -- number of entries is bounded by the number of base urls times the number
-    -- of chains.
-    --
-    -- We could add a counter that would reap the map from stall streams every
-    -- nth time getUpdateVar.
-
-    -- We don't restart streams automatically, in order to save connections.
-    -- Thus there is a risk that a stream dies while a mining loop is subscribed
-    -- to it. We solve this by preempting the loop if we haven't seen an update
-    -- after 2 times the block time.
-
-withPreemption :: UpdateMap -> UpdateKey -> RIO Env a -> RIO Env (Either () a)
-withPreemption m k inner = do
+-- | Run an operation that is preempted if an update event occurs.
+--
+-- Streams are restarted automatically, when they got closed by the server. We
+-- don't restart streams automatically in case of a failure, but instead throw
+-- an exception. Failures are supposed to be handled in the outer mining
+-- functions.
+--
+-- There risk that a stream stalls without explicitely failing. We solve this by
+-- preempting the loop if we haven't seen an update after 5 times the block time
+-- (which will affect about 0.7% of all blocks).
+--
+withPreemption :: UpdateKey -> RIO Env a -> RIO Env (Either () a)
+withPreemption k inner = do
+    m <- asks envUpdateMap
     var <- getUpdateVar m k
     race (awaitChange var) inner
   where
     awaitChange var = do
         cur <- readTVarIO var
-        timeoutVar <- registerDelay 60000000 -- 1 minute -- FIXME this should be 2*block time
+        timeoutVar <- registerDelay (5 * 30_000_000) -- 5 times the block time ~ 0.7% of all blocks
         atomically $ do
             isTimeout <- readTVar timeoutVar
             isUpdate <- (/= cur) <$> readTVar var
             unless (isTimeout || isUpdate) retrySTM
 
+-- | Atomatically restarts the stream when the response status is 2** and throws
+-- and exception otherwise.
+--
 updateStream :: ChainId -> TVar Int -> RIO Env ()
 updateStream cid var = do
     e <- ask
     u <- NEL.head <$> readIORef (envUrls e) -- Do we ever use something else than the head?
-    liftIO $ withEvents (req u) (envMgr e) $ \updates -> updates
+    forever $ liftIO $ withEvents (req u) (envMgr e) $ \updates -> updates
```

Comment:
I see that in [in 735](https://github.com/kadena-io/chainweb-node/pull/735/files#diff-22f2620e094b259e8e42309f48831467R174) you're returning a `CloseEvent`, but this isn't checked for in `realEvent`. Is this intentional?

Reply:
yes, after `CloseEvent` the server will terminate the HTTP request and `withEvents` will close the stream and return.

Reply:
And that's automatic within `wai` ?

Reply:
yes, wai does the right thing here. The question is wether `withEvents` terminates the response properly. I think, I checked and the parse will just terminate and exit the stream when it sees the end of the input stream.